### PR TITLE
JDK-8283730: Improve discussion of modeling of packages and modules

### PR DESCRIPTION
**Pull request by jddarcy**
Informative update explicitly linking PackageElement objects to package-info.java files and ModuleElement objects to module-info.java files.

I'll reflow the updated paragraph in javax.lang.model.element.package-info before pushing.

Plain text update to PackageElement:

The represented package may have an explicit source code or executable output backing construct or may be created from implicit information. The explicit source code construct for a package is typically a package-info.java file (JLS 7.4.1). Implicit information is used to model unnamed packages as well as named packages without explicit declarations.

Plain text update to ModuleElement:

The represented module may have an explicit source code or executable output backing construct or may be created from implicit information. The explicit source code construct for a module is typically a module-info.java file (JLS 7.7). Implicit information is used to model unnamed modules.

Plain text update to TypeElement (added for symmetry with above updates):

The represented class or interface may have an explicit source code or executable output backing construct. Multiple classes and interfaces can share the same backing construct. For example, multiple classes and interface can be declared in the same source file, including, but are not limited to, a top-level class or interface and auxiliary classes and interfaces or a top-level class or interface and nested classes within it.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283730](https://bugs.openjdk.java.net/browse/JDK-8283730): Improve discussion of modeling of packages and modules

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to bc93fe17dbbf862ba519421e91817c8b1295cfb6
 * [Jan Lahoda](https://openjdk.java.net/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to a2944516015d3dc48cdc2b00435736c7946f3ef7

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7980/head:pull/7980` \
`$ git checkout pull/7980`

Update a local copy of the PR: \
`$ git checkout pull/7980` \
`$ git pull https://git.openjdk.java.net/jdk pull/7980/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7980`

View PR using the GUI difftool: \
`$ git pr show -t 7980`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7980.diff">https://git.openjdk.java.net/jdk/pull/7980.diff</a>

</details>

### CODE REVIEW

File: src/java.compiler/share/classes/javax/lang/model/element/TypeElement.java [hunk 1]
```diff
@@ -53,6 +53,25 @@
  * source of information is Java source code, then the elements will be
  * returned in source code order.
  *
+ * @apiNote
+ * The represented class or interface may have an explicit backing
+ * construct (either {@linkplain
+ * javax.annotation.processing.Filer#createSourceFile(CharSequence,
```
Comment (lahodaj) [inline src/java.compiler/share/classes/javax/lang/model/element/TypeElement.java:59]: The `Filer` references here feel a bit weird. While most `TypeElement`s presumably have an explicit backing construct (i.e. a file), I suspect that only minority of the backing constructs are created using a `Filer`, but the text suggests(?) that all `TypeElement`s with a backing construct are created via `Filer`.

Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/element/TypeElement.java]: In the latest pushed, pulled out the information specific to an annotation processing. New plain text:

ModuleElement:
"The represented module may have an explicit reference representation (either source code or executable output) or may be created from implicit information. The explicit and standalone source code construct for a module is typically a module-info.java file (JLS 7.7). Automatic modules (JLS 7.7.1) are named modules that do not have a module-info file. Implicit information is used to model unnamed modules.

In the context of annotation processing, a module element can be:
-     created from the initial inputs to a run of the tool
-     queried for in the configured environment 

PackageElement
The represented package may have an explicit reference representation (either source code or executable output) or may be created from implicit information. The explicit and standalone source code construct for a package is typically a package-info.java file (JLS 7.4.1). Implicit information is used to model unnamed packages.

In the context of annotation processing, a package element can be:
-     created from the initial inputs to a run of the tool
-     created from source code or class files written by a processor
-     queried for in the configured environment 

TypeElement
    The represented class or interface may have a reference representation (either source code or executable output). Multiple classes and interfaces can share the same reference representation backing construct. For example, multiple classes and interface can be declared in the same source file, including, but are not limited to:
-         a top-level class or interface and auxiliary classes and interfaces
-         a top-level class or interface and nested class and interfaces within it 

    In the context of annotation processing, a type element can be:
-         created from the initial inputs to a run of the tool
-         created from source code or class files written by a processor
-         queried for in the configured environment

File: src/java.compiler/share/classes/javax/lang/model/element/PackageElement.java [hunk 2]
```diff
@@ -32,6 +32,28 @@
...
  * about the package and its members.
  *
+ * @apiNote
+ * The represented package may have an explicit {@linkplain
+ * javax.lang.model.util.Elements#getFileObjectOf(Element) reference
+ * representation} (either source code or executable output) or may be
+ * created from implicit information. The explicit and standalone
+ * source code construct for a package is typically a {@code
+ * package-info.java} file (JLS {@jls 7.4.1}). Implicit information is
+ * used to model {@linkplain #isUnnamed unnamed packages}.
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/lang/model/element/PackageElement.java:42]: you mention unnamed packages but not named packages without `package-info.java`

Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/element/PackageElement.java]: Added discussion on that point; will reflow paragraphs later.

**Comment by jddarcy** [comment]
> _Mailing list message from [Alex Buckley](mailto:alex.buckley@oracle.com) on [compiler-dev](mailto:compiler-dev@mail.openjdk.java.net):_
> 
> On 3/30/2022 6:49 PM, Joseph D. Darcy wrote:
> 
> > On 3/29/2022 9:18 AM, Alex Buckley wrote:
> > > On 3/28/2022 6:30 PM, Joseph D. Darcy wrote:
> > > > "The represented module may have an explicit backing construct
> > > > (either source code or executable output) or may be created from
> > > > implicit information. The explicit and standalone source code
> > > > construct for a module is typically a module-info.java file (JLS
> > > > 7.7). Implicit information is used to model unnamed modules and
> > > > may be used for automatic modules (JLS 7.7.1)."
> > 
> > 
> > In the full javadoc rendering of the text in question, "automatic
> > modules" ... is a link to the utility method
> > javax.lang.model.util.Elements.isAutomaticModule, from which reader
> > could infer that ...
> > Consistent with other aspect of the design of javax.lang.model, ...
> 
> OK. However, I have strong reservations about whether the subclause "and standalone" is effective at communicating what we want. The factoring is not quite right -- a better factoring would be "The explicit source code construct for a module is typically a standalone module-info.java file". And for PackageElement: "The explicit source code construct for a package is typically a standalone package-info.java file."
> 
> Alex

Latest update to the PackageElement note:

The represented package may have an explicit reference representation (either source code or executable output) or may be created from implicit information. The explicit and standalone source code construct for a package is typically a package-info.java file (JLS 7.4.1). A named package without a standalone package-info.java file can be declared by the package declarations in a top-level class or interface. Implicit information is used to model unnamed packages (JLS 7.4.2).

In the context of annotation processing, a package element can be:
- created from the initial inputs to a run of the tool
- created from source code or class files written by a processor
- queried for in the configured environment

File: src/java.compiler/share/classes/javax/lang/model/element/PackageElement.java [hunk 3]
```diff
@@ -32,6 +32,33 @@
...
+
+ * The represented package may have an explicit {@linkplain
+ * javax.lang.model.util.Elements#getFileObjectOf(Element) reference
+ * representation} (either source code or executable output) or may be
+ * created from implicit information. The explicit and standalone
+ * source code construct for a package is typically a {@code
+ * package-info.java} file (JLS {@jls 7.4.1}).
+ * A named package without a standalone {@code package-info.java} file
+ * can be declared by the package declarations in a {@linkplain
+ * NestingKind#TOP_LEVEL top-level} class or interface.
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/lang/model/element/PackageElement.java:45]: Grammar pedantry:  
1. There are not "package declarations" (plural) in a (singular) top-level class or interface.
2. I'd also quibble whether the package declaration is "in" a top-level class or interface. Maybe you can rephrase using "of" or "for".

**Review by jonathan-gibbons — APPROVED** [review]
I've posted some grammar quibbles for you, but otherwise approved it.

File: src/java.compiler/share/classes/javax/lang/model/element/PackageElement.java (hunk 3 above)
Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/element/PackageElement.java:45]: Will fix point 1.

JLS contains the sentence: "A package declaration in an ordinary compilation unit specifies the name of the package to which the compilation unit belongs."

So I'll change to "A named package without ... can be declared in the package declaration of a top-level class or interface."

Thanks.

**Pull request closed without merging**